src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Transmission/OtlpExporterPersistentStorageRetryTransmissionHandler.cs

```diff
@@ -0,0 +1,110 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#nullable enable
+
+using Google.Protobuf;
+using OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.ExportClient;
+using OpenTelemetry.PersistentStorage.Abstractions;
+using OpenTelemetry.PersistentStorage.FileSystem;
+using OpenTelemetry.Proto.Collector.Logs.V1;
+using OpenTelemetry.Proto.Collector.Metrics.V1;
+using OpenTelemetry.Proto.Collector.Trace.V1;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.Transmission;
+
+internal sealed class OtlpExporterPersistentStorageRetryTransmissionHandler<TRequest> : OtlpExporterTransmissionHandler<TRequest>
+{
+    private readonly ManualResetEvent stopEvent = new(false);
+    private readonly Thread thread;
+    private readonly PersistentBlobProvider? persistentBlobProvider;
+    private readonly Func<byte[], TRequest>? requestFactory;
+
+    public OtlpExporterPersistentStorageRetryTransmissionHandler(IExportClient<TRequest> exportClient, Func<byte[], TRequest> requestFactory, string storagePath)
+        : base(exportClient)
+    {
+        this.persistentBlobProvider = new FileBlobProvider(storagePath);
+        this.requestFactory = requestFactory;
+
+        this.thread = new Thread(this.ProcessRetryRequestsThreadBody)
+        {
+            Name = $"OtlpExporter Persistent Retry Storage - {typeof(TRequest)}",
+            IsBackground = true,
+        };
+
+        this.thread.Start();
+    }
+
+    protected override bool OnSubmitRequestFailure(TRequest request, ExportClientResponse response)
+    {
+        byte[]? data = null;
```

Comment:
Could we have a big buffer stored on a threadstatic or something for this?

Reply:
Will have to check. .ToByteArray() extension from Google.Protobuf always creates a new array.

Reply:
👍

On a sort of related note, I'm wondering if we should wrap the byte[] returned from protobuf with our own envelope. Primarily to introduce a version byte/indicator at the front. I have found introducing a version byte into binary blobs to be _very_ useful thing in the past. For example if we bump protobuf and it suddenly starts writing things a different way we could use that version to discard things or convert them. Or perhaps in the future we move away from protobuf to custom serialization or whatever. Having a byte/indicator in there can be a life saver when deploying new code and starting up with old files 😄